File: code/cpp/kth-largest-element-in-an-array.cc
// 9 ms, 09/26/2017
// Time  : O(n lg k)
// Space : O(k)
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k)
    {
        std::priority_queue<int, std::vector<int>, std::greater<int>> heap;
        for (auto x : nums)
            if (heap.size() < k || heap.top() < x) {
                heap.push(x);
                if (heap.size() > k)
                    heap.pop();
            }
        return heap.top();
    }
};
/*---------------------------------------------------------------------------*/
// Cleaner code
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k)
    {
        std::priority_queue<int, std::vector<int>, std::greater<int>> heap(nums.begin(), nums.begin() + k);
        for (int i = k, n = nums.size(); i < n; ++i)
            if (heap.top() < nums[i]) {
                heap.push(nums[i]);
                heap.pop();
            }
        return heap.top();
    }
};
/*---------------------------------------------------------------------------*/
// Linear solution
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k)
    {
        nth_element(nums.begin(), nums.begin() + k - 1, nums.end(), std::greater<int>());
        return nums[k - 1];
    }
};

File: code/python/kth-largest-element-in-an-array.py
# 02/19/2018
# Time  : O(n*lgk)
# Space : O(n)

class Solution:
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        
        # Standard technique using a head
        #  - maintain a list of k largest elements using a min-heap
        #  - return the top of the heap at the end
        
        import heapq
        min_heap = []
        for x in nums:
            heapq.heappush(min_heap, x)
            if len(min_heap) > k: # maintain min_heap size of k
                heapq.heappop(min_heap)
        return min_heap[0]
